src/annotator/highlighter.ts

```diff
@@ -21,6 +21,195 @@ export const clusterValues: HighlightCluster[] = [
+
+      const parent = nodes[0].parentNode as ParentNode;
+      parent.replaceChild(highlightEl, nodes[0]);
+      nodes.forEach(node => highlightEl.appendChild(node));
+
+      highlights.push(highlightEl);
+    });
+
+    // For PDF highlights, create the highlight effect by using an SVG placed
+    // above the page's canvas rather than CSS `background-color` on the highlight
+    // element. This enables more control over blending of the highlight with the
+    // content below.
+    //
+    // Drawing these SVG highlights involves measuring the `<hypothesis-highlight>`
+    // elements, so we create them only after those elements have all been created
+    // to reduce the number of forced reflows. We also skip creating them for
+    // unrendered pages for performance reasons.
+    if (!inPlaceholder) {
+      drawHighlightsAbovePDFCanvas(highlights, cssClass);
+    }
+
+    return highlights;
+  }
+
+  /**
+   * Remove all highlights under a given root element.
+   */
+  removeAllHighlights() {
+    const highlights = Array.from(
+      this.root.querySelectorAll('hypothesis-highlight'),
+    );
+    this.removeHighlights(highlights as HighlightElement[]);
+  }
+
+  /**
+   * Remove highlights from a range previously highlighted with `highlightRange`.
+   */
+  removeHighlights(highlights: HighlightElement[]) {
+    // Explicitly un-focus highlights to be removed. This ensures associated
+    // focused elements are removed from the document.
+    setHighlightsFocused(highlights, false);
+    for (const h of highlights) {
+      if (h.parentNode) {
+        const children = Array.from(h.childNodes);
+        replaceWith(h, children);
+      }
+      if (h.svgHighlight) {
+        h.svgHighlight.remove();
+      }
+    }
+  }
+
+  /**
+   * Set whether the given highlight elements should appear "focused".
+   *
+   * A highlight can be displayed in a different ("focused") style to indicate
+   * that it is current in some other context - for example the user has selected
+   * the corresponding annotation in the sidebar.
+   */
+  setHighlightsFocused(highlights: HighlightElement[], focused: boolean) {
+    highlights.forEach(h => {
+      // In PDFs the visible highlight is created by an SVG element, so the focused
+      // effect is applied to that. In other documents the effect is applied to the
+      // `<hypothesis-highlight>` element.
+      if (h.svgHighlight) {
+        setSVGHighlightFocused(h.svgHighlight, focused);
+      } else {
+        h.classList.toggle('hypothesis-highlight-focused', focused);
+      }
+    });
+  }
+
+  /**
+   * Set whether highlights under the given root element should be visible.
+   */
+  setHighlightsVisible(visible: boolean) {
+    this.root.classList.toggle(showHighlightsClass, visible);
+  }
+
+  /**
+   * Get the visible highlight elements at the given client coordinates.
+   */
+  getHighlightsFromPoint(x: number, y: number): HighlightElement[] {
+    return getHighlightsFromPoint(x, y);
```

Comment:
`getHighlightsFromPoint` is still exposed as a standalone function because the PDF integration uses it, and doesn't have access to the highlighter yet. That's an exercise for a future PR.